### PR TITLE
fix(redis): use sentinel mode and ACL username in prod clients

### PR DESCRIPTION
**Pull request by glopez-dev**
## Summary

- Introduce `buildRedisConnection()` factory in `src/modules/app/redis-connection.ts` that reads `REDIS_MODE` and branches between `direct` and `sentinel`, always passing `username`, `password`, `db`, and (in sentinel mode) `sentinelPassword`. Also sets `reconnectOnError` on `NOAUTH` so the client self-heals after a credential rotation.
- Wire the factory into every Redis consumer: BullMQ (`app.module.ts`), the custom health service, the monitoring controller, and the cache-manager (switching `@keyv/redis` to `createSentinel` in sentinel mode so it uses `@redis/client`'s sentinel support).
- Unit-test the factory (direct + sentinel, missing-env guards, NOAUTH reconnect).

## Why

In `whispr-prod`, `scheduling-service` pods were spamming `WRONGPASS` / `NOAUTH` errors every few seconds. Redis there is configured with ACL auth **and** sentinel discovery, and the `Deployment` already injects `REDIS_MODE=sentinel`, `REDIS_USERNAME`, `REDIS_SENTINELS`, `REDIS_MASTER_NAME`, and `REDIS_SENTINEL_PASSWORD`. But every ioredis client was built inline with only `{ host, port, password, db }`, ignoring the username (breaking ACL auth) and ignoring sentinel discovery (no failover). BullMQ had the same bug silently, meaning scheduled messages were plausibly never being dispatched in prod.

This aligns the service with the pattern already used in `auth-service/src/config/redis.config.ts`.

## Test plan

- [x] Unit tests green (`npx jest src/modules/app/redis-connection.spec.ts`)
- [x] Lint clean on touched files
- [ ] E2E tests — broken on main by a preexisting `@nestjs/jwt` type resolution issue unrelated to this PR; CI runs e2e in Docker so it should be fine there
- [ ] After merge + deploy: verify `WRONGPASS` / `NOAUTH` errors disappear from `whispr-prod` pod logs
- [ ] After merge + deploy: verify a scheduled message created via the API actually fires at its scheduled time

Closes WHISPR-807

### CODE REVIEW

File: src/modules/app/redis-connection.ts [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+import { RedisOptions } from 'ioredis';
+
+export function parseSentinels(sentinelsStr: string): Array<{ host: string; port: number }> {
+	return sentinelsStr
+		.split(',')
+		.map((s) => s.trim())
+		.filter((s) => s.length > 0)
+		.map((s) => {
+			const [host, port] = s.split(':');
+			return { host, port: Number.parseInt(port, 10) };
```
[the remark this answers is not in the document]
Comment (glopez-dev) [reply src/modules/app/redis-connection.ts]: Fixed in 7adb47e. Added validation for host:port format, non-empty host, and finite integer port. Also added 3 unit tests covering the invalid-entry paths.

File: src/modules/app/redis-connection.ts [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+export function buildRedisConnection(configService: ConfigService): RedisOptions {
+	const mode = configService.get<string>('REDIS_MODE', 'direct');
+	const db = Number.parseInt(configService.get<string>('REDIS_DB', '0'), 10);
+	const username = configService.get<string>('REDIS_USERNAME') || undefined;
+	const password = configService.get<string>('REDIS_PASSWORD') || undefined;
+
+	const reconnectOnError = (err: Error) => err.message.includes('NOAUTH');
+
+	if (mode === 'sentinel') {
+		const sentinelsStr = configService.get<string>('REDIS_SENTINELS');
```
[the remark this answers is not in the document]
Comment (glopez-dev) [reply src/modules/app/redis-connection.ts]: Fixed in 7adb47e. REDIS_MODE is now validated against an allowlist (`direct` | `sentinel`) and throws for unsupported values. Unit test added.

File: src/modules/app/redis-connection.ts [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+			username,
+			password,
+			sentinelPassword,
+			reconnectOnError,
+		};
+	}
+
+	const host = configService.get<string>('REDIS_HOST', 'localhost');
+	const port = Number.parseInt(configService.get<string>('REDIS_PORT', '6379'), 10);
+
```
[the remark this answers is not in the document]
Comment (glopez-dev) [reply src/modules/app/redis-connection.ts:94]: Fixed in 7adb47e. Both `REDIS_DB` and `REDIS_PORT` are now validated with `Number.isFinite()` and throw a descriptive error when non-numeric. Unit tests added for both.

File: src/modules/app/redis-connection.ts [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+		});
+}
+
+export function buildRedisConnection(configService: ConfigService): RedisOptions {
+	const mode = configService.get<string>('REDIS_MODE', 'direct');
+	const db = Number.parseInt(configService.get<string>('REDIS_DB', '0'), 10);
+	const username = configService.get<string>('REDIS_USERNAME') || undefined;
+	const password = configService.get<string>('REDIS_PASSWORD') || undefined;
+
+	const reconnectOnError = (err: Error) => err.message.includes('NOAUTH');
```
[the remark this answers is not in the document]
Comment (glopez-dev) [reply src/modules/app/redis-connection.ts]: Fixed in 7adb47e. `reconnectOnError` now matches both `NOAUTH` and `WRONGPASS`. Updated the existing test to cover both cases.

File: src/modules/app/cache.ts [hunk 5]
```diff
@@ -1,14 +1,50 @@
...
+
+		if (!sentinelsStr) {
+			throw new Error('REDIS_SENTINELS is required when REDIS_MODE=sentinel');
+		}
+		if (!masterName) {
+			throw new Error('REDIS_MASTER_NAME is required when REDIS_MODE=sentinel');
+		}
+		if (!sentinelPassword) {
+			throw new Error('REDIS_SENTINEL_PASSWORD is required when REDIS_MODE=sentinel');
+		}
```
[the remark this answers is not in the document]
Comment (glopez-dev) [reply src/modules/app/cache.ts:24]: Acknowledged. The duplication exists because `cache.ts` uses `@keyv/redis` (`createClient`/`createSentinel`) which has a different option shape than ioredis (`buildRedisConnection`). The two clients cannot share a single factory without an adapter layer, which would add complexity disproportionate to the benefit for this PR. I'll note it as a follow-up for when we consolidate Redis clients.

File: src/modules/app/redis-connection.ts [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+		.map((s) => {
+			const [host, port] = s.split(':');
+			return { host, port: Number.parseInt(port, 10) };
+		});
+}
+
+export function buildRedisConnection(configService: ConfigService): RedisOptions {
+	const mode = configService.get<string>('REDIS_MODE', 'direct');
+	const db = Number.parseInt(configService.get<string>('REDIS_DB', '0'), 10);
+	const username = configService.get<string>('REDIS_USERNAME') || undefined;
```
[the remark this answers is not in the document]
Comment (glopez-dev) [reply src/modules/app/redis-connection.ts:44]: Fixed in 7adb47e — same as the earlier thread: `REDIS_DB` and `REDIS_PORT` are now validated with `Number.isFinite()`.

File: src/modules/app/cache.ts [hunk 7]
```diff
@@ -1,14 +1,50 @@
...
+
+		client = createSentinel({
+			name: masterName,
+			sentinelRootNodes: parseSentinels(sentinelsStr),
+			nodeClientOptions: { username, password, database: db },
+			sentinelClientOptions: { password: sentinelPassword },
+		});
```
[the remark this answers is not in the document]
Comment (glopez-dev) [reply src/modules/app/cache.ts]: Acknowledged. Adding unit tests for `cacheModuleOptionsFactory` requires mocking `@keyv/redis` internals (`createClient`/`createSentinel`) which is non-trivial and out of scope for this bug fix PR. The sentinel env-var guards are identical to those already tested via `buildRedisConnection`. I'll track cache factory tests as a follow-up.

File: src/modules/monitoring/services/health.service.ts [hunk 8]
```diff
@@ -19,17 +20,19 @@ export class CustomHealthService extends HealthIndicator {
 		private configService: ConfigService
 	) {
 		super();
+		const { reconnectOnError: _, ...connOpts } = buildRedisConnection(this.configService);
```
[the remark this answers is not in the document]
Comment (glopez-dev) [reply src/modules/monitoring/services/health.service.ts]: Fixed in 7adb47e. Removed the destructuring — `reconnectOnError` is now passed through as intended by the factory. The original exclusion was a workaround for an e2e hang caused by a separate issue (CI Redis having no password while `REDIS_PASSWORD` was set in test env), which has been fixed at the source.

File: src/modules/monitoring/controllers/monitoring.controller.ts [hunk 9]
```diff
@@ -35,17 +37,19 @@ export class MonitoringController {
 		private readonly dataSource: DataSource,
 		private readonly configService: ConfigService
 	) {
+		const { reconnectOnError: _, ...connOpts } = buildRedisConnection(this.configService);
```
[the remark this answers is not in the document]
Comment (glopez-dev) [reply src/modules/monitoring/controllers/monitoring.controller.ts]: Fixed in 7adb47e. Same as the health service thread — removed the destructuring, `reconnectOnError` is now passed through. The underlying e2e issue was fixed in `setup-e2e.ts`.

File: src/modules/app/redis-connection.ts [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+		};
+	}
+
+	const host = configService.get<string>('REDIS_HOST', 'localhost');
+	const port = Number.parseInt(configService.get<string>('REDIS_PORT', '6379'), 10);
+
+	return {
+		host,
+		port,
+		db,
```
[the remark this answers is not in the document]
Comment (glopez-dev) [reply src/modules/app/redis-connection.ts:98]: Fixed in 7adb47e — same as earlier threads: `REDIS_PORT` is now validated with `Number.isFinite()` and throws a descriptive error.

File: src/modules/app/redis-connection.ts [hunk 11]
```diff
@@ -0,0 +1,82 @@
...
+
+export function buildRedisConnection(configService: ConfigService): RedisOptions {
+	const mode = configService.get<string>('REDIS_MODE', 'direct');
+	if (mode !== 'direct' && mode !== 'sentinel') {
+		throw new Error(`Unsupported REDIS_MODE "${mode}": must be "direct" or "sentinel"`);
+	}
+	const db = Number.parseInt(configService.get<string>('REDIS_DB', '0'), 10);
+	if (!Number.isFinite(db)) {
+		throw new Error(`Invalid REDIS_DB: must be a finite integer`);
+	}
```
[the remark this answers is not in the document]
Comment (glopez-dev) [reply src/modules/app/redis-connection.ts]: Fixed in e5fb442. Switched from `Number.parseInt()` to `Number()` + `Number.isInteger()` — this now rejects partially-numeric inputs like `"0abc"`. Also enforces `REDIS_DB >= 0` and `REDIS_PORT` in range 1–65535. Added tests for both edge cases.

File: src/modules/app/cache.ts [hunk 12]
```diff
@@ -1,14 +1,50 @@
...
 import { ConfigService } from '@nestjs/config';
-import KeyvRedis from '@keyv/redis';
+import KeyvRedis, { createClient, createSentinel } from '@keyv/redis';
+import { parseSentinels } from './redis-connection';
 
 export function cacheModuleOptionsFactory(configService: ConfigService): CacheOptions {
-	const redis_host = configService.get('REDIS_HOST', 'redis');
-	const redis_port = configService.get('REDIS_PORT', 6379);
-	const redis_url = `redis://${redis_host}:${redis_port}`;
+	const mode = configService.get<string>('REDIS_MODE', 'direct');
```
[the remark this answers is not in the document]
Comment (glopez-dev) [reply src/modules/app/cache.ts]: Fixed in e5fb442. `cacheModuleOptionsFactory` now validates `REDIS_MODE` against the same allowlist (`direct` | `sentinel`) and throws for unsupported values, consistent with `buildRedisConnection`.

File: src/modules/app/cache.ts [hunk 13]
```diff
@@ -1,14 +1,50 @@
...
+
+		client = createSentinel({
+			name: masterName,
+			sentinelRootNodes: parseSentinels(sentinelsStr),
+			nodeClientOptions: { username, password, database: db },
+			sentinelClientOptions: { password: sentinelPassword },
+		});
+	} else {
+		const host = configService.get<string>('REDIS_HOST', 'redis');
+		const port = Number.parseInt(configService.get<string>('REDIS_PORT', '6379'), 10);
```
[the remark this answers is not in the document]
Comment (glopez-dev) [reply src/modules/app/cache.ts]: Fixed in e5fb442. `REDIS_PORT` in `cache.ts` is now validated with `Number()` + `Number.isInteger()` and range check (1–65535), same as `buildRedisConnection`.

File: src/modules/app/cache.ts [hunk 14]
```diff
@@ -1,14 +1,50 @@
...
-import KeyvRedis from '@keyv/redis';
+import KeyvRedis, { createClient, createSentinel } from '@keyv/redis';
+import { parseSentinels } from './redis-connection';
 
 export function cacheModuleOptionsFactory(configService: ConfigService): CacheOptions {
-	const redis_host = configService.get('REDIS_HOST', 'redis');
-	const redis_port = configService.get('REDIS_PORT', 6379);
-	const redis_url = `redis://${redis_host}:${redis_port}`;
+	const mode = configService.get<string>('REDIS_MODE', 'direct');
+	const db = Number.parseInt(configService.get<string>('REDIS_DB', '0'), 10);
```
[the remark this answers is not in the document]
Comment (glopez-dev) [reply src/modules/app/cache.ts]: Fixed in e5fb442. `REDIS_DB` in `cache.ts` is now validated with `Number()` + `Number.isInteger()` and `>= 0` check, same as `buildRedisConnection`.

File: src/modules/app/redis-connection.ts [hunk 15]
```diff
@@ -0,0 +1,84 @@
...
+			}
+			const host = parts[0].trim();
+			const portStr = parts[1].trim();
+			if (host.length === 0) {
+				throw new Error(`Invalid REDIS_SENTINELS entry "${s}": host must be non-empty`);
+			}
+			const port = Number.parseInt(portStr, 10);
+			if (!Number.isFinite(port)) {
+				throw new Error(`Invalid REDIS_SENTINELS entry "${s}": port must be a finite integer`);
+			}
```
[the remark this answers is not in the document]
Comment (glopez-dev) [reply src/modules/app/redis-connection.ts]: Fixed in 0b54108. Switched from `Number.parseInt()` to `Number()` + `Number.isInteger()` for sentinel port parsing — this now rejects partially-numeric inputs like `"26379abc"`. Also added 1–65535 range validation. Tests added for partially-numeric port, out-of-range port, and port 0.

File: src/modules/app/redis-connection.ts [hunk 16]
```diff
@@ -0,0 +1,84 @@
...
+		}
+		if (!masterName) {
+			throw new Error('REDIS_MASTER_NAME is required when REDIS_MODE=sentinel');
+		}
+		if (!sentinelPassword) {
+			throw new Error('REDIS_SENTINEL_PASSWORD is required when REDIS_MODE=sentinel');
+		}
+
+		return {
+			sentinels: parseSentinels(sentinelsStr),
```
[the remark this answers is not in the document]
Comment (glopez-dev) [reply src/modules/app/redis-connection.ts]: Fixed in 0b54108. Both `buildRedisConnection()` and `cacheModuleOptionsFactory()` now check that `parseSentinels()` returns at least one entry before proceeding, throwing a targeted error for inputs like `",,"`. Added a unit test covering this case.
